src/oom/oomd-manager.c

```diff
@@ -0,0 +1,439 @@
+        STRV_FOREACH(p, paths) {
+                _cleanup_free_ char *key = NULL;
+
+                r = recursively_get_cgroup_context(m->monitored_cgroup_contexts_children, new_children, *p);
+                if (r == -ENOMEM)
+                        return log_oom();
+
+                /* We don't want the cgroups we're watching in the children hashmap so remove them
+                 * and put them in a separate hashmap for detection only. */
+                ctx = hashmap_remove2(new_children, *p, (void **) &key);
+                if (ctx) {
+                        r = hashmap_put(new_base, TAKE_PTR(key), ctx);
+                        if (r == -ENOMEM)
+                                return log_oom();
+                }
+        }
+
+        hashmap_free_free_free(m->monitored_cgroup_contexts_children);
+        m->monitored_cgroup_contexts_children = TAKE_PTR(new_children);
+        hashmap_free_free_free(m->monitored_cgroup_contexts);
+        m->monitored_cgroup_contexts = TAKE_PTR(new_base);
+
+        // TODO anitazha clean up this debugging mess
+        OomdCGroupContext *c;
+        Iterator i;
+        char *key;
+        HASHMAP_FOREACH_KEY(c, key, m->monitored_cgroup_contexts, i) {
+                log_info("me: %s", key);
+        }
+
+        HASHMAP_FOREACH_KEY(c, key, m->monitored_cgroup_contexts_children, i) {
+                log_info("children: %s", key);
+        }
+
+        return 0;
+}
+
+static int monitor_cgroup_contexts_handler(sd_event_source *s, uint64_t usec, void *userdata) {
+        Manager *m = userdata;
+        const char *target;
+        bool swap_trigger;
+        usec_t usec_now;
+        int r;
+
+        assert(s);
+        assert(userdata);
+
+        /* Reset timer */
+        r = sd_event_now(sd_event_source_get_event(s), CLOCK_MONOTONIC, &usec_now);
+        if (r < 0)
+                return r;
+
+        r = sd_event_source_set_time(s, usec_now + INTERVAL_USEC);
+        if (r < 0)
+                return r;
+
+        /* Update the watch values */
+        r = update_monitored_cgroup_contexts(m);
+        if (r == -ENOMEM)
+                return log_oom();
+
+        r = oomd_system_context_get("/proc/swaps", &m->system_context);
+        if (r == -ENOMEM)
+                return log_oom();
+
+        /* If we're still recovering from a kill, don't try to kill again yet */
+        if (m->post_action_delay_start + POST_ACTION_DELAY_USEC >= usec_now)
+                m->post_action_delay_start = 0;
+        else
+                return 0;
+
+        /* Check if we need to kill things */
+        if (FLAGS_SET(m->detection_mode, RESOURCE_METRIC_MASK_MEM_PRESSURE)) {
+                target = oomd_pressure_above(m->monitored_cgroup_contexts,
+                                             PRESSURE_THRESHOLD,
+                                             PRESSURE_DURATION_USEC,
+                                             &m->last_hit_mem_thres);
```

Comment:
You likely want to also check that there has been some reclaim activity on the cgroup in the last interval (pgscan > 0). The concern is the following case: Pressure climbed, a lot of high-frequency pages were reclaimed, we killed the offending cgroup. Even after this, well-behaved processes will fault in recently resident pages and this will cause pressure to remain high.

The post_action_delay helps you a bit here, but better to address it more directly - if you aren't seeing any reclaim pressure, no need to kill something (it won't help anyways).